File: intellij-plugin/src/main/kotlin/com/enterprise/aiplugin/mcp/McpClient.kt
package com.enterprise.aiplugin.mcp

import com.intellij.openapi.diagnostic.Logger
import kotlinx.coroutines.*
import java.net.URI
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import java.net.http.HttpClient
import java.net.http.WebSocket
import java.net.http.WebSocket.Listener
import java.util.concurrent.CompletionStage

data class McpTool(
    val name: String,
    val description: String,
    val parameters: List<McpParameter>,
    val returns: McpParameter,
    var enabled: Boolean = true,
    val permissions: List<String> = emptyList()
)

data class McpParameter(
    val name: String,
    val type: String,
    val description: String,
    val required: Boolean,
    val default: Any? = null
)

data class McpToolCall(
    val tool: String,
    val parameters: Map<String, Any>,
    val id: String
)

data class McpToolResult(
    val id: String,
    val result: Any?,
    val error: String? = null,
    val executionTime: Long = 0
)

data class McpServerConfig(
    val name: String,
    val url: String,
    var enabled: Boolean = false,
    val permissions: List<String> = emptyList(),
    val tools: List<String> = emptyList()
)

class McpClient {
    private val logger = Logger.getInstance(McpClient::class.java)
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    private val servers = ConcurrentHashMap<String, McpServerConfig>()
    private val tools = ConcurrentHashMap<String, McpTool>()
    private val connections = ConcurrentHashMap<String, WebSocket>()
    private val toolRegistry = ConcurrentHashMap<String, List<McpTool>>()
    private val pendingResults = ConcurrentHashMap<String, CompletableFuture<McpToolResult>>()
    private val callIdCounter = AtomicInteger(0)
    private val httpClient = HttpClient.newHttpClient()
    
    private var isConnected = false

    init {
        loadDefaultServers()
    }

    /**
     * Connect to MCP servers
     */
    suspend fun connect(): Unit = withContext(Dispatchers.IO) {
        try {
            for ((serverName, config) in servers) {
                if (config.enabled) {
                    connectToServer(serverName, config)
                }
            }
            isConnected = true
            logger.info("Connected to MCP servers")
        } catch (error: Exception) {
            logger.error("Failed to connect to MCP servers", error)
            throw error
        }
    }

    /**
     * Connect to a specific MCP server
     */
    private suspend fun connectToServer(serverName: String, config: McpServerConfig): Unit = withContext(Dispatchers.IO) {
        try {
            val uri = URI(config.url)
            
            httpClient.newWebSocketBuilder()
                .buildAsync(uri, object : Listener {
                    override fun onOpen(webSocket: WebSocket) {
                        logger.info("Connected to MCP server: $serverName")
                        connections[serverName] = webSocket
                        
                        // Discover tools
                        scope.launch {
                            discoverTools(serverName)
                        }
                    }

                    override fun onClose(webSocket: WebSocket, statusCode: Int, reason: String): CompletionStage<*>? {
                        logger.info("Disconnected from MCP server: $serverName")
                        connections.remove(serverName)
                        return null
                    }

                    override fun onError(webSocket: WebSocket, error: Throwable) {
                        logger.error("MCP server error ($serverName)", error)
                    }

                    override fun onText(webSocket: WebSocket, data: CharSequence, last: Boolean): CompletionStage<*>? {
                        handleServerMessage(serverName, data.toString())
                        return null
                    }
                }).join()

        } catch (error: Exception) {
            logger.error("Failed to connect to MCP server: $serverName", error)
            throw error
        }
    }

    /**
     * Discover available tools from a server
     */
    private suspend fun discoverTools(serverName: String): Unit = withContext(Dispatchers.IO) {
        val connection = connections[serverName] ?: return@withContext

        val discoveryMessage = mapOf(
            "type" to "tool_discovery",
            "server" to serverName,
            "timestamp" to System.currentTimeMillis()
        )

        connection.sendText(discoveryMessage.toJson(), true)
    }

    /**
     * Handle messages from MCP servers
     */
    private fun handleServerMessage(serverName: String, message: String) {
        try {
            val data = message.fromJson<Map<String, Any>>()
            
            when (data["type"]) {
                "tool_list" -> {
                    val toolsList = (data["tools"] as? List<Map<String, Any>>)?.map { toolData ->
                        McpTool(
                            name = toolData["name"] as String,
                            description = toolData["description"] as String,
                            parameters = (toolData["parameters"] as? List<Map<String, Any>>)?.map { param ->
                                McpParameter(
                                    name = param["name"] as String,
                                    type = param["type"] as String,
                                    description = param["description"] as String,
                                    required = param["required"] as Boolean,
                                    default = param["default"]
                                )
                            } ?: emptyList(),
                            returns = McpParameter(
                                name = "result",
                                type = "any",
                                description = "Tool result",
                                required = true
                            ),
                            enabled = toolData["enabled"] as? Boolean ?: true,
                            permissions = (toolData["permissions"] as? List<String>) ?: emptyList()
                        )
                    } ?: emptyList()
                    
                    registerTools(serverName, toolsList)
                }
                
                "tool_result" -> {
                    val result = McpToolResult(
                        id = data["id"] as String,
                        result = data["result"],
                        error = data["error"] as? String,
                        executionTime = (data["executionTime"] as? Number)?.toLong() ?: 0
                    )
                    
                    handleToolResult(result)
                }
                
                "error" -> {
                    logger.error("MCP server error ($serverName): ${data["error"]}")
                }
                
                else -> {
                    logger.warn("Unknown MCP message type: ${data["type"]}")
                }
            }
        } catch (error: Exception) {
            logger.error("Failed to parse MCP message", error)
        }
    }

    /**
     * Register tools from a server
     */
    private fun registerTools(serverName: String, toolsList: List<McpTool>) {
        val serverTools = mutableListOf<McpTool>()
        
        for (tool in toolsList) {
            val toolId = "$serverName:${tool.name}"
            tools[toolId] = tool
            serverTools.add(tool)
        }
        
        toolRegistry[serverName] = serverTools
        logger.info("Registered ${toolsList.size} tools from server: $serverName")
    }

    /**
     * Call a tool
     */
    suspend fun callTool(toolName: String, parameters: Map<String, Any>): McpToolResult = withContext(Dispatchers.IO) {
        val tool = findTool(toolName)
        if (tool == null) {
            throw IllegalArgumentException("Tool not found: $toolName")
        }

        if (!tool.enabled) {
            throw IllegalStateException("Tool is disabled: $toolName")
        }

        val serverName = getServerForTool(toolName)
        val connection = connections[serverName]
        
        if (connection == null) {
            throw IllegalStateException("Server not connected: $serverName")
        }

        val callId = generateCallId()

        val toolCall = McpToolCall(
            tool = toolName,
            parameters = parameters,
            id = callId
        )

        val message = mapOf(
            "type" to "tool_call",
            "call" to toolCall,
            "timestamp" to System.currentTimeMillis()
        )

        val future = CompletableFuture<McpToolResult>()
        pendingResults[callId] = future

        try {
            // Send tool call
            connection.sendText(message.toJson(), true)

            // Wait for result with timeout
            val result = future.get(30, TimeUnit.SECONDS)
            result
        } catch (error: Exception) {
            pendingResults.remove(callId)
            throw RuntimeException("Tool call timeout or error: $toolName", error)
        }
    }

    /**
     * Handle tool result
     */
    private fun handleToolResult(result: McpToolResult) {
        val future = pendingResults.remove(result.id)
        if (future != null) {
            future.complete(result)
        }
    }

    /**
     * Get available tools
     */
    fun getAvailableTools(): List<McpTool> {
        return tools.values.toList()
    }

    /**
     * Get tools by server
     */
    fun getToolsByServer(serverName: String): List<McpTool> {
        return toolRegistry[serverName] ?: emptyList()
    }

    /**
     * Enable/disable a tool
     */
    fun setToolEnabled(toolName: String, enabled: Boolean) {
        val tool = findTool(toolName)
        if (tool != null) {
            tool.enabled = enabled
        }
    }

    /**
     * Add a custom tool
     */
    fun addCustomTool(tool: McpTool, serverName: String = "custom") {
        val toolId = "$serverName:${tool.name}"
        tools[toolId] = tool
        
        val serverTools = toolRegistry.getOrDefault(serverName, emptyList()).toMutableList()
        serverTools.add(tool)
        toolRegistry[serverName] = serverTools
    }

    /**
     * Remove a tool
     */
    fun removeTool(toolName: String) {
        val tool = findTool(toolName)
        if (tool != null) {
            val serverName = getServerForTool(toolName)
            tools.remove("$serverName:$toolName")
            
            val serverTools = toolRegistry[serverName]?.toMutableList()
            if (serverTools != null) {
                serverTools.removeAll { it.name == toolName }
                toolRegistry[serverName] = serverTools
            }
        }
    }

    /**
     * Add MCP server configuration
     */
    fun addServer(config: McpServerConfig) {
        servers[config.name] = config
    }

    /**
     * Remove MCP server
     */
    fun removeServer(serverName: String) {
        servers.remove(serverName)
        
        // Close connection if exists
        val connection = connections[serverName]
        if (connection != null) {
            try {
                connection.sendClose(WebSocket.NORMAL_CLOSURE, "Removing server")
            } catch (error: Exception) {
                logger.error("Error closing connection to server: $serverName", error)
            }
            connections.remove(serverName)
        }
        
        // Remove tools from this server
        val serverTools = toolRegistry[serverName] ?: emptyList()
        for (tool in serverTools) {
            tools.remove("$serverName:${tool.name}")
        }
        toolRegistry.remove(serverName)
    }

    /**
     * Get server configuration
     */
    fun getServerConfig(serverName: String): McpServerConfig? {
        return servers[serverName]
    }

    /**
     * Get all server configurations
     */
    fun getAllServerConfigs(): List<McpServerConfig> {
        return servers.values.toList()
    }

    /**
     * Check if connected
     */
    fun isConnectedToServers(): Boolean {
        return isConnected && connections.isNotEmpty()
    }

    /**
     * Disconnect from all servers
     */
    fun disconnect() {
        for ((serverName, connection) in connections) {
            try {
                connection.sendClose(WebSocket.NORMAL_CLOSURE, "Disconnecting")
            } catch (error: Exception) {
                logger.error("Error closing connection to server: $serverName", error)
            }
        }
        connections.clear()
        isConnected = false
    }

    // Helper methods
    private fun findTool(toolName: String): McpTool? {
        return tools.values.find { it.name == toolName }
    }

    private fun getServerForTool(toolName: String): String {
        for ((id, tool) in tools) {
            if (tool.name == toolName) {
                return id.split(":")[0]
            }
        }
        throw IllegalArgumentException("Tool not found: $toolName")
    }

    private fun generateCallId(): String {
        return "call_${System.currentTimeMillis()}_${callIdCounter.incrementAndGet()}"
    }

    private fun loadDefaultServers() {
        // Add some default MCP server configurations
        val defaultServers = listOf(
            McpServerConfig(
                name = "local-tools",
                url = "ws://localhost:8080/mcp",
                enabled = false,
                permissions = listOf("read", "write"),
                tools = emptyList()
            ),
            McpServerConfig(
                name = "enterprise-tools",
                url = "wss://tools.enterprise-ai-plugin.com/mcp",
                enabled = false,
                permissions = listOf("read"),
                tools = emptyList()
            )
        )

        for (server in defaultServers) {
            servers[server.name] = server
        }
    }

    // JSON serialization helpers
    private fun Map<String, Any>.toJson(): String {
        return com.google.gson.Gson().toJson(this)
    }

    private inline fun <reified T> String.fromJson(): T {
        return com.google.gson.Gson().fromJson(this, T::class.java)
    }
}

File: vs-code-plugin/src/mcp/McpClient.ts
import * as vscode from 'vscode';
import WebSocket from 'ws';

export interface McpTool {
    name: string;
    description: string;
    parameters: McpParameter[];
    returns: McpParameter;
    enabled: boolean;
    permissions: string[];
}

export interface McpParameter {
    name: string;
    type: string;
    description: string;
    required: boolean;
    default?: any;
}

export interface McpToolCall {
    tool: string;
    parameters: any;
    id: string;
}

export interface McpToolResult {
    id: string;
    result: any;
    error?: string;
    executionTime: number;
}

export interface McpServerConfig {
    name: string;
    url: string;
    enabled: boolean;
    permissions: string[];
    tools: string[];
}

export class McpClient {
    private servers: Map<string, McpServerConfig> = new Map();
    private tools: Map<string, McpTool> = new Map();
    private connections: Map<string, WebSocket> = new Map();
    private isConnected: boolean = false;
    private toolRegistry: Map<string, McpTool[]> = new Map();

    constructor() {
        this.loadDefaultServers();
    }

    /**
     * Connect to MCP servers
     */
    public async connect(): Promise<void> {
        try {
            for (const [serverName, config] of this.servers) {
                if (config.enabled) {
                    await this.connectToServer(serverName, config);
                }
            }
            this.isConnected = true;
        } catch (error) {
            throw new Error(`Failed to connect to MCP servers: ${error}`);
        }
    }

    /**
     * Connect to a specific MCP server
     */
    private async connectToServer(serverName: string, config: McpServerConfig): Promise<void> {
        return new Promise((resolve, reject) => {
            try {
                const ws = new WebSocket(config.url);
                
                ws.on('open', () => {
                    console.log(`Connected to MCP server: ${serverName}`);
                    this.connections.set(serverName, ws);
                    
                    // Discover tools
                    this.discoverTools(serverName);
                    resolve();
                });

                ws.on('message', (data: WebSocket.Data) => {
                    this.handleServerMessage(serverName, data);
                });

                ws.on('error', (error) => {
                    console.error(`MCP server error (${serverName}):`, error);
                    reject(error);
                });

                ws.on('close', () => {
                    console.log(`Disconnected from MCP server: ${serverName}`);
                    this.connections.delete(serverName);
                });

            } catch (error) {
                reject(error);
            }
        });
    }

    /**
     * Discover available tools from a server
     */
    private async discoverTools(serverName: string): Promise<void> {
        const connection = this.connections.get(serverName);
        if (!connection) return;

        const discoveryMessage = {
            type: 'tool_discovery',
            server: serverName,
            timestamp: Date.now()
        };

        connection.send(JSON.stringify(discoveryMessage));
    }

    /**
     * Handle messages from MCP servers
     */
    private handleServerMessage(serverName: string, data: WebSocket.Data): void {
        try {
            const message = JSON.parse(data.toString());
            
            switch (message.type) {
                case 'tool_list':
                    this.registerTools(serverName, message.tools);
                    break;
                    
                case 'tool_result':
                    this.handleToolResult(message);
                    break;
                    
                case 'error':
                    console.error(`MCP server error (${serverName}):`, message.error);
                    break;
                    
                default:
                    console.log(`Unknown MCP message type: ${message.type}`);
            }
        } catch (error) {
            console.error(`Failed to parse MCP message:`, error);
        }
    }

    /**
     * Register tools from a server
     */
    private registerTools(serverName: string, tools: McpTool[]): void {
        const serverTools: McpTool[] = [];
        
        for (const tool of tools) {
            const toolId = `${serverName}:${tool.name}`;
            this.tools.set(toolId, tool);
            serverTools.push(tool);
        }
        
        this.toolRegistry.set(serverName, serverTools);
        console.log(`Registered ${tools.length} tools from server: ${serverName}`);
    }

    /**
     * Call a tool
     */
    public async callTool(toolName: string, parameters: any): Promise<McpToolResult> {
        const tool = this.findTool(toolName);
        if (!tool) {
            throw new Error(`Tool not found: ${toolName}`);
        }

        if (!tool.enabled) {
            throw new Error(`Tool is disabled: ${toolName}`);
        }

        const serverName = this.getServerForTool(toolName);
        const connection = this.connections.get(serverName);
        
        if (!connection) {
            throw new Error(`Server not connected: ${serverName}`);
        }

        return new Promise((resolve, reject) => {
            const callId = this.generateCallId();
            const startTime = Date.now();

            const toolCall: McpToolCall = {
                tool: toolName,
                parameters,
                id: callId
            };

            const message = {
                type: 'tool_call',
                call: toolCall,
                timestamp: Date.now()
            };

            // Set up result handler
            const resultHandler = (result: McpToolResult) => {
                if (result.id === callId) {
                    resolve(result);
                }
            };

            // Store result handler (simplified - in real implementation you'd have a proper event system)
            (this as any).resultHandlers = (this as any).resultHandlers || new Map();
            (this as any).resultHandlers.set(callId, resultHandler);

            // Send tool call
            connection.send(JSON.stringify(message));

            // Timeout after 30 seconds
            setTimeout(() => {
                reject(new Error(`Tool call timeout: ${toolName}`));
            }, 30000);
        });
    }

    /**
     * Handle tool result
     */
    private handleToolResult(message: any): void {
        const result: McpToolResult = {
            id: message.id,
            result: message.result,
            error: message.error,
            executionTime: message.executionTime || 0
        };

        // Find and call result handler
        const resultHandlers = (this as any).resultHandlers;
        if (resultHandlers && resultHandlers.has(result.id)) {
            const handler = resultHandlers.get(result.id);
            handler(result);
            resultHandlers.delete(result.id);
        }
    }

    /**
     * Get available tools
     */
    public getAvailableTools(): McpTool[] {
        return Array.from(this.tools.values());
    }

    /**
     * Get tools by server
     */
    public getToolsByServer(serverName: string): McpTool[] {
        return this.toolRegistry.get(serverName) || [];
    }

    /**
     * Enable/disable a tool
     */
    public setToolEnabled(toolName: string, enabled: boolean): void {
        const tool = this.findTool(toolName);
        if (tool) {
            tool.enabled = enabled;
        }
    }

    /**
     * Add a custom tool
     */
    public addCustomTool(tool: McpTool, serverName: string = 'custom'): void {
        const toolId = `${serverName}:${tool.name}`;
        this.tools.set(toolId, tool);
        
        if (!this.toolRegistry.has(serverName)) {
            this.toolRegistry.set(serverName, []);
        }
        this.toolRegistry.get(serverName)!.push(tool);
    }

    /**
     * Remove a tool
     */
    public removeTool(toolName: string): void {
        const tool = this.findTool(toolName);
        if (tool) {
            const serverName = this.getServerForTool(toolName);
            this.tools.delete(`${serverName}:${toolName}`);
            
            const serverTools = this.toolRegistry.get(serverName);
            if (serverTools) {
                const index = serverTools.findIndex(t => t.name === toolName);
                if (index !== -1) {
                    serverTools.splice(index, 1);
                }
            }
        }
    }

    /**
     * Add MCP server configuration
     */
    public addServer(config: McpServerConfig): void {
        this.servers.set(config.name, config);
    }

    /**
     * Remove MCP server
     */
    public removeServer(serverName: string): void {
        this.servers.delete(serverName);
        
        // Close connection if exists
        const connection = this.connections.get(serverName);
        if (connection) {
            connection.close();
            this.connections.delete(serverName);
        }
        
        // Remove tools from this server
        const serverTools = this.toolRegistry.get(serverName) || [];
        for (const tool of serverTools) {
            this.tools.delete(`${serverName}:${tool.name}`);
        }
        this.toolRegistry.delete(serverName);
    }

    /**
     * Get server configuration
     */
    public getServerConfig(serverName: string): McpServerConfig | undefined {
        return this.servers.get(serverName);
    }

    /**
     * Get all server configurations
     */
    public getAllServerConfigs(): McpServerConfig[] {
        return Array.from(this.servers.values());
    }

    /**
     * Check if connected
     */
    public isConnectedToServers(): boolean {
        return this.isConnected && this.connections.size > 0;
    }

    /**
     * Disconnect from all servers
     */
    public disconnect(): void {
        for (const [serverName, connection] of this.connections) {
            connection.close();
        }
        this.connections.clear();
        this.isConnected = false;
    }

    // Helper methods
    private findTool(toolName: string): McpTool | undefined {
        for (const [id, tool] of this.tools) {
            if (tool.name === toolName) {
                return tool;
            }
        }
        return undefined;
    }

    private getServerForTool(toolName: string): string {
        for (const [id, tool] of this.tools) {
            if (tool.name === toolName) {
                return id.split(':')[0];
            }
        }
        throw new Error(`Tool not found: ${toolName}`);
    }

    private generateCallId(): string {
        return `call_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }

    private loadDefaultServers(): void {
        // Add some default MCP server configurations
        const defaultServers: McpServerConfig[] = [
            {
                name: 'local-tools',
                url: 'ws://localhost:8080/mcp',
                enabled: false,
                permissions: ['read', 'write'],
                tools: []
            },
            {
                name: 'enterprise-tools',
                url: 'wss://tools.enterprise-ai-plugin.com/mcp',
                enabled: false,
                permissions: ['read'],
                tools: []
            }
        ];

        for (const server of defaultServers) {
            this.servers.set(server.name, server);
        }
    }
}
